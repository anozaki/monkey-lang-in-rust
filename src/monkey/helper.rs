pub fn is_identifier(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

pub fn is_digit(ch: char) -> bool {
    ch.is_ascii_digit()
}

pub fn is_whitespace(ch: char) -> bool {
    ch.is_whitespace()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_is_identifier() {
        assert_eq!(is_identifier('a'), true);
        assert_eq!(is_identifier('_'), true);
        assert_eq!(is_identifier('あ'), false);
        assert_eq!(is_identifier('1'), false);
    }

    #[test]
    fn test_is_digit() {
        assert_eq!(is_digit('1'), true);
        assert_eq!(is_digit('2'), true);
        assert_eq!(is_digit('3'), true);
        assert_eq!(is_digit('4'), true);
        assert_eq!(is_digit('_'), false);
        assert_eq!(is_digit('%'), false);
        assert_eq!(is_digit('１'), false); // non ascii 1
    }

    #[test]
    fn test_is_whitespace() {
        assert_eq!(is_whitespace(' '), true);
        assert_eq!(is_whitespace('\n'), true);
        assert_eq!(is_whitespace('\t'), true);
        assert_eq!(is_whitespace('　'), true); // non-ascii space
    }
}