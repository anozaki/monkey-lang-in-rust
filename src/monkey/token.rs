use std::fmt::{Debug, Formatter};

#[derive(Debug, Copy, Clone, PartialEq)]
pub enum TokenType {
    Illegal,
    EndOfFile,

    Comment,

    String,

    // Identifier/Literals
    Identifier,
    Integer,

    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,

    LessThan,
    GreaterThan,

    Equal,
    NotEqual,
    LessThanEqual,
    GreaterThanEqual,

    // Delimiters
    Comma,
    Colon,
    Semicolon,

    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,

    // Keywords
    Function,
    Let,
    True,
    False,
    If,
    Else,
    Return,
}

#[derive(PartialEq, Clone)]
pub struct Span {
    pub start: usize,
    pub end: usize,
    pub line_start: usize,
    pub line_end: usize,
}

impl Debug for Span {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f, "[L{}-{:?}:{:?}]",
            self.line_start, self.start, self.end
        )
    }
}

#[derive(PartialEq, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,

    pub span: Span,
}

impl Debug for Token {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f, "Token({:?}, {:?} {:?})",
            self.token_type, self.literal, self.span
        )
    }
}
