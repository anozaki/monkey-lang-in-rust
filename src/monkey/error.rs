use std::rc::Rc;
use thiserror::Error;
use crate::monkey::token::Token;

#[derive(Error, Debug, PartialEq, Clone)]
pub enum Error {
    #[error("Generic {0}")]
    TodoError(String),

    #[error("Invalid token found at {0:?}")]
    InvalidTokenError(Rc<Token>),

    #[error("Invalid escape char {0:?}")]
    InvalidCharError(char),
}