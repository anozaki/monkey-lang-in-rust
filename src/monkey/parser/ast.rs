use std::collections::HashMap;
use std::fmt::{Debug, Display, Formatter};
use std::rc::Rc;

use crate::monkey::token::Token;

#[derive(Debug, Clone, PartialEq)]
pub enum Operator {
    Not,
    Neg,
    Add,
    Sub,
    Mul,
    Div,
    Greater,
    Less,
    Equal,
    NotEqual,
    Call,
    Index,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Program {
    pub statements: Vec<StatementNode>,
}


#[derive(Debug, Clone, PartialEq)]
pub struct Identifier(pub String);

#[derive(Debug, Clone, PartialEq)]
pub enum StatementNode {
    Let(Identifier, Box<ExpressionNode>),
    Return(Box<ExpressionNode>),
    If {
        condition: Box<ExpressionNode>,
        consequence: Box<Program>,
        alternative: Option<Box<Program>>,
    },
    Expression {
        expression: Box<ExpressionNode>
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionNode {
    Identifier(Identifier),
    Bool(bool),
    Int(isize),
    String(String),
    Call {
        function: Box<ExpressionNode>,
        params: Vec<Box<ExpressionNode>>,
    },
    Function {
        params: Vec<Identifier>,
        body: Box<Program>,
    },
    Infix {
        operator: Operator,
        left: Box<ExpressionNode>,
        right: Box<ExpressionNode>,
    },
    Prefix {
        operator: Operator,
        expression: Box<ExpressionNode>,
    },
    ArrayLiteral {
        params: Vec<Box<ExpressionNode>>,
    },
    Index {
        left: Box<ExpressionNode>,
        index: Box<ExpressionNode>,
    },
    HashLiteral {
        params: Vec<(Box<ExpressionNode>, Box<ExpressionNode>)>,
    },
}

pub struct Node {
    token: Rc<Token>,
    program: Program,
}
