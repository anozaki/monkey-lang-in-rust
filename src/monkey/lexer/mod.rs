use std::cell::Cell;
use std::collections::VecDeque;
use std::fs;
use std::rc::Rc;
use crate::monkey::error::Error::{InvalidTokenError, TodoError, InvalidCharError};
use crate::monkey::Result;

use crate::monkey::helper::{is_digit, is_identifier, is_whitespace};
use crate::monkey::token::{Span, Token, TokenType};

pub trait Lexer {
    fn token(&self) -> Result<Token>;
}

pub struct MonkeyLexer {
    /// Input
    input: Vec<char>,

    column: Cell<usize>,
    position: Cell<usize>,
    read_position: Cell<usize>,
    line: Cell<usize>,
    ch: Cell<char>,
}

impl Lexer for MonkeyLexer {
    fn token(&self) -> Result<Token> {
        self.read()
    }
}


impl MonkeyLexer {
    pub fn new(input: &str) -> Self {
        let input: Vec<char> = input.chars().collect();
        let ch: char = input[0];
        Self {
            input,
            column: Cell::new(0),
            position: Cell::new(0),
            read_position: Cell::new(1),
            line: Cell::new(0),
            ch: Cell::new(ch.into()),
        }
    }

    fn peek(&self) -> char {
        self.input[self.read_position.get()]
    }

    fn handle_whitespace(&self) {
        while is_whitespace(self.ch.get()) {
            self.next()
        }
    }

    fn handle_single_token(&self, token_type: TokenType) -> Result<Token> {
        let token = self.tokenize(1, 0, token_type);
        self.next();

        token
    }

    fn handle_double_token(&self, token_type: TokenType) -> Result<Token> {
        let token = self.tokenize(2, 0, token_type);
        self.next();
        self.next();

        token
    }

    fn read_operator_double(&self, token_type: TokenType) -> Result<Token> {
        if self.peek() == '=' {
            match self.ch.get() {
                '=' => self.handle_double_token(TokenType::Equal),
                '!' => self.handle_double_token(TokenType::NotEqual),
                '<' => self.handle_double_token(TokenType::GreaterThanEqual),
                '>' => self.handle_double_token(TokenType::LessThanEqual),
                c @ _ => Err(TodoError(format!("unsupported token {}", c))),
            }
        } else {
            self.handle_single_token(token_type)
        }
    }

    fn read(&self) -> Result<Token> {
        self.handle_whitespace();
        match self.ch.get() {
            '=' => self.read_operator_double(TokenType::Assign),
            '+' => self.handle_single_token(TokenType::Plus),
            '-' => self.handle_single_token(TokenType::Minus),
            '!' => self.read_operator_double(TokenType::Bang),
            '*' => self.handle_single_token(TokenType::Asterisk),
            '/' => self.handle_single_token(TokenType::Slash),
            '<' => self.read_operator_double(TokenType::LessThan),
            '>' => self.read_operator_double(TokenType::GreaterThan),
            ';' => self.handle_single_token(TokenType::Semicolon),
            '(' => self.handle_single_token(TokenType::LeftParen),
            ')' => self.handle_single_token(TokenType::RightParen),
            ',' => self.handle_single_token(TokenType::Comma),
            '{' => self.handle_single_token(TokenType::LeftBrace),
            '}' => self.handle_single_token(TokenType::RightBrace),
            '[' => self.handle_single_token(TokenType::LeftBracket),
            ']' => self.handle_single_token(TokenType::RightBracket),
            ':' => self.handle_single_token(TokenType::Colon),
            '"' => self.handle_string_token(),
            '\0' => self.tokenize(0, 0, TokenType::EndOfFile),
            ch if is_identifier(ch) => self.read_identifier(),
            ch if is_digit(ch) => self.read_digit(),
            c @ _ => Err(TodoError(format!("unsupported token {}", c))),
        }
    }

    fn read_identifier(&self) -> Result<Token> {
        let position = self.position.get();
        while is_identifier(self.ch.get()) {
            self.next();
        }
        let end_position = self.position.get();
        let length = end_position - position;

        let identifier: &str = &self.input[position..end_position].iter().collect::<String>();
        match identifier {
            "let" => self.tokenize(length, length, TokenType::Let),
            "fn" => self.tokenize(length, length, TokenType::Function),
            "true" => self.tokenize(length, length, TokenType::True),
            "false" => self.tokenize(length, length, TokenType::False),
            "if" => self.tokenize(length, length, TokenType::If),
            "else" => self.tokenize(length, length, TokenType::Else),
            "return" => self.tokenize(length, length, TokenType::Return),
            _ => self.tokenize(length, length, TokenType::Identifier),
        }
    }

    fn read_digit(&self) -> Result<Token> {
        let position = self.position.get();
        while is_digit(self.ch.get()) {
            self.next()
        }
        let end_position = self.position.get();
        let length = end_position - position;

        self.tokenize(length, length, TokenType::Integer)
    }

    /// Move the processing position to the next token
    fn next(&self) {
        if self.input.len() <= self.read_position.get() {
            self.ch.set(0.into());
            self.column.set(self.column.get() + 1);
        } else {
            if self.input[self.position.get()] == '\n' {
                self.line.set(self.line.get() + 1);
                self.column.set(0)
            } else {
                self.column.set(self.column.get() + 1);
            }
            self.ch.set(self.input[self.read_position.get()])
        }

        self.position.set(self.position.get() + 1);
        self.read_position.set(self.read_position.get() + 1);
    }

    fn handle_string_token(&self) -> Result<Token> {
        let position = self.position.get();
        let start = self.column.get();

        self.next();

        let mut chars: Vec<char> = Vec::new();

        while self.ch.get() != '"' && self.ch.get() != '\0' {
            if self.ch.get() == '\\' {
                self.next();
                match self.ch.get() {
                    'n' => chars.push('\n'),
                    't' => chars.push('\t'),
                    '"' => chars.push('"'),
                    '\\' => chars.push('\\'),
                    val @ _ => return Err(InvalidCharError(self.ch.get())),
                }
            } else {
                chars.push(self.ch.get())
            }
            self.next()
        }

        let end_position = self.position.get();
        let length = end_position - position;
        let end = self.column.get();

        self.next();

        // FIXME: there is some code duplication here - some how merge this with tokenizer.
        Ok(Token {
            token_type: TokenType::String,
            literal: chars.iter().collect::<String>(),
            span: Span {
                start,
                end,
                line_start: self.line.get(),
                line_end: self.line.get(),
            },
        })
    }

    fn tokenize(&self, length: usize, offset: usize, token_type: TokenType) -> Result<Token> {
        let position = self.position.get() - offset;

        let literal = if length > 0 {
            self.input[position..position + length].to_vec()
        } else {
            vec![]
        };

        let column = self.column.get() - offset;
        let end = if length > 0 { column + length - 1 } else { column };
        Ok(Token {
            token_type,
            literal: literal.iter().collect::<String>(),
            span: Span {
                start: column,
                end,
                line_start: self.line.get(),
                line_end: self.line.get(),
            },
        })
    }
}

pub fn run_lexer(source: &str) -> Result<String> {
    let contents = fs::read_to_string(source).unwrap();
    eval_lexer(&contents)
}

pub fn eval_lexer(contents: &String) -> Result<String> {
    let lines: Vec<usize> = contents
        .chars()
        .enumerate()
        .filter(|(_, ch)| *ch == '\n')
        .map(|(idx, _)| idx)
        .collect();

    let lexer = MonkeyLexer::new(&*contents);

    let mut tokens = VecDeque::new();
    let mut stop_loop = false;
    while !stop_loop {
        let result = lexer.token();
        if let Ok(token) = result {
            stop_loop = token.token_type == TokenType::EndOfFile;
            tokens.push_back(token);
        } else if let Err(message) = result {
            return Err(message);
        }
    }

    let snapshot = token_snapshot(&contents, &lines, &tokens);

    Ok(snapshot)
}

pub fn token_snapshot(program: &String, lines: &Vec<usize>, tokens: &VecDeque<Token>) -> String {
    let mut output = String::new();
    let mut working_line = usize::MAX;
    for token in tokens {
        if token.span.line_start != token.span.line_end {
            panic!("multiline command not supported ðŸ˜ž")
        }

        if token.span.line_start != working_line {
            working_line = token.span.line_start;
            let start = if working_line != 0 {
                lines[working_line - 1]
            } else {
                0
            };

            let end = if working_line < lines.len() {
                lines[working_line]
            } else {
                program.len() - 1
            };
            output += &program[start..=end];
        }

        output += &" ".repeat(token.span.start);
        output += &"^".repeat(token.span.end + 1 - token.span.start);
        output += &format!(" {token:?}");
        output += &"\n";
    }

    output
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_lexer() {
        let result = run_lexer("monkey/lexer.mky").unwrap();

        insta::assert_snapshot!(result)
    }
}
