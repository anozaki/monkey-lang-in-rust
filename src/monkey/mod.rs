use crate::monkey::error::Error;

pub mod error;

pub mod lexer;
pub mod helper;
pub mod parser;
pub mod token;


pub mod interpreter;

pub type Result<T> = core::result::Result<T, Error>;
