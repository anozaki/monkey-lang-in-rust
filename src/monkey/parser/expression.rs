use std::collections::HashMap;
use std::rc::Rc;

use crate::monkey::error::Error;
use crate::monkey::parser::ast::{ExpressionNode, Identifier, Operator, StatementNode};
use crate::monkey::parser::Parser;
use crate::monkey::Result;
use crate::monkey::token::{Token, TokenType};
use crate::try_next;

#[derive(Copy, Clone)]
pub enum OrderOfOps {
    Invalid,
    Lowest,
    Equals,
    LessGreater,
    Sum,
    Product,
    Prefix,
    Call,
    Index,
}

impl<'a> Parser<'a> {
    pub fn parse_expression(&mut self) -> Result<StatementNode> {
        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        Ok(StatementNode::Expression { expression })
    }

    fn order_of_operation(&self, token_type: TokenType) -> OrderOfOps {
        match token_type {
            TokenType::Equal | TokenType::NotEqual => OrderOfOps::Equals,
            TokenType::LessThan | TokenType::GreaterThan => OrderOfOps::LessGreater,
            TokenType::Plus | TokenType::Minus => OrderOfOps::Sum,
            TokenType::Asterisk | TokenType::Slash => OrderOfOps::Product,
            TokenType::LeftParen => OrderOfOps::Call,
            TokenType::LeftBracket => OrderOfOps::Index,
            _ => OrderOfOps::Invalid,
        }
    }
    pub fn parse_expression_node(&mut self, precedence: OrderOfOps) -> Result<Box<ExpressionNode>> {
        let mut left_result: Box<ExpressionNode> = match self.current().token_type {
            TokenType::Identifier => self.parse_identifier()?,
            TokenType::Integer => self.parse_integer_literal()?,
            TokenType::Bang => self.parse_prefix(Operator::Not)?,
            TokenType::Minus => self.parse_prefix(Operator::Neg)?,
            TokenType::True => self.parse_bool()?,
            TokenType::False => self.parse_bool()?,
            TokenType::LeftParen => self.parse_group()?,
            TokenType::Function => self.parse_fn()?,
            TokenType::String => self.parse_string()?,
            TokenType::LeftBracket => self.parse_array_literal()?,
            TokenType::LeftBrace => self.parse_hash_literal()?,
            _ => return Err(Error::InvalidTokenError(Rc::clone(self.current())))
        };

        let precedence_order = precedence as isize;
        while self.peek().token_type != TokenType::Semicolon && precedence_order < self.order_of_operation(self.peek().token_type) as isize {
            self.next();

            let infix = self.parse_infix_node(left_result)?;

            left_result = infix;
        }

        Ok(left_result)
    }

    pub fn parse_identifier(&self) -> Result<Box<ExpressionNode>> {
        let result = ExpressionNode::Identifier(Identifier(self.current().literal.clone()));
        Ok(Box::new(result))
    }

    fn parse_infix_node(&mut self, left: Box<ExpressionNode>) -> Result<Box<ExpressionNode>> {
        Ok(match self.current().token_type {
            TokenType::Plus => self.parse_infix(left, Operator::Add)?,
            TokenType::Minus => self.parse_infix(left, Operator::Sub)?,
            TokenType::Slash => self.parse_infix(left, Operator::Div)?,
            TokenType::Asterisk => self.parse_infix(left, Operator::Mul)?,
            TokenType::Equal => self.parse_infix(left, Operator::Equal)?,
            TokenType::NotEqual => self.parse_infix(left, Operator::NotEqual)?,
            TokenType::LessThan => self.parse_infix(left, Operator::Less)?,
            TokenType::GreaterThan => self.parse_infix(left, Operator::Greater)?,
            TokenType::LeftParen => self.parse_call(left, Operator::Call)?,
            TokenType::LeftBracket => self.parse_index(left, Operator::Index)?,
            token @ _ => return Err(Error::InvalidTokenError(Rc::clone(self.current())))
        })
    }

    fn parse_prefix(&mut self, operator: Operator) -> Result<Box<ExpressionNode>> {
        let token: Rc<Token> = Rc::clone(self.current());

        self.next();

        let expression = self.parse_expression_node(OrderOfOps::Prefix)?;

        Ok(Box::new(ExpressionNode::Prefix { operator, expression }))
    }

    fn parse_infix(&mut self, left: Box<ExpressionNode>, operator: Operator) -> Result<Box<ExpressionNode>> {
        let token: Rc<Token> = Rc::clone(self.current());
        let precedence = self.order_of_operation(self.current().token_type);

        self.next();

        let right = self.parse_expression_node(precedence)?;

        Ok(Box::new(ExpressionNode::Infix { operator, left, right }))
    }

    fn parse_group(&mut self) -> Result<Box<ExpressionNode>> {
        let token: Rc<Token> = Rc::clone(self.current());

        self.next();

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        Ok(expression)
    }


    fn parse_fn(&mut self) -> Result<Box<ExpressionNode>> {
        let token = Rc::clone(&self.token_current);
        self.next();

        try_next!(self, TokenType::LeftParen);
        let params = self.parse_fn_param()?;
        try_next!(self, TokenType::RightParen);

        try_next!(self, TokenType::LeftBrace);
        let body = self.parse_program()?;

        if self.peek().token_type == TokenType::Semicolon {
            try_next!(self, TokenType::RightBrace);
        }

        Ok(Box::new(ExpressionNode::Function { params, body }))
    }

    fn parse_fn_param(&mut self) -> Result<Vec<Identifier>> {
        let mut result: Vec<Identifier> = Vec::new();

        if self.current().token_type == TokenType::RightParen {
            return Ok(result);
        }

        loop {
            result.push(Identifier(self.current().literal.clone()));
            self.next();

            if self.current().token_type == TokenType::RightParen {
                break;
            }
            try_next!(self, TokenType::Comma);
        }

        Ok(result)
    }
    fn parse_call(&mut self, function: Box<ExpressionNode>, _: Operator) -> Result<Box<ExpressionNode>> {
        let token = Rc::clone(&self.token_current);

        try_next!(self, TokenType::LeftParen);
        let params = self.parse_expression_list(TokenType::RightParen)?;
        // try_next!(self, TokenType::RightParen);

        Ok(Box::new(ExpressionNode::Call { function, params }))
    }

    fn parse_index(&mut self, left: Box<ExpressionNode>, _: Operator) -> Result<Box<ExpressionNode>> {
        let token = Rc::clone(&self.token_current);

        try_next!(self, TokenType::LeftBracket);
        let index = self.parse_expression_node(OrderOfOps::Lowest)?;
        self.next();

        return Ok(Box::new(
            ExpressionNode::Index { left, index }
        ));
    }

    fn parse_expression_list(&mut self, end_token: TokenType) -> Result<Vec<Box<ExpressionNode>>> {
        let mut result: Vec<Box<ExpressionNode>> = Vec::new();

        if self.current().token_type == end_token {
            return Ok(result);
        }

        loop {
            result.push(self.parse_expression_node(OrderOfOps::Lowest)?);
            self.next();

            if self.current().token_type == end_token {
                break;
            }
            try_next!(self, TokenType::Comma);
        }

        Ok(result)
    }

    pub fn parse_integer_literal(&mut self) -> Result<Box<ExpressionNode>> {
        let Ok(value) = self.current().literal.parse::<isize>() else {
            return Err(Error::InvalidTokenError(Rc::clone(self.current())));
        };

        Ok(Box::new(ExpressionNode::Int(value)))
    }

    pub fn parse_bool(&mut self) -> Result<Box<ExpressionNode>> {
        let value = match self.current().token_type {
            TokenType::True => true,
            TokenType::False => false,
            _ => false,
        };

        Ok(Box::new(ExpressionNode::Bool(value)))
    }
    fn parse_string(&self) -> Result<Box<ExpressionNode>> {
        Ok(Box::new(
            ExpressionNode::String(
                self.current().literal.clone()
            )
        ))
    }

    fn parse_array_literal(&mut self) -> Result<Box<ExpressionNode>> {
        let token = Rc::clone(&self.token_current);

        try_next!(self, TokenType::LeftBracket);
        let params = self.parse_expression_list(TokenType::RightBracket)?;

        Ok(Box::new(ExpressionNode::ArrayLiteral { params }))
    }

    fn parse_hash_literal(&mut self) -> Result<Box<ExpressionNode>> {
        let mut params: Vec<(Box<ExpressionNode>, Box<ExpressionNode>)> = Vec::new();

        if self.current().token_type == TokenType::RightBrace {
            return Ok(Box::new(ExpressionNode::HashLiteral { params }));
        }

        try_next!(self, TokenType::LeftBrace);
        loop {
            let key = self.parse_expression_node(OrderOfOps::Lowest)?;
            self.next();


            try_next!(self, TokenType::Colon);
            let val = self.parse_expression_node(OrderOfOps::Lowest)?;
            self.next();

            params.push((key, val));

            if self.current().token_type == TokenType::RightBrace {
                break;
            }
            try_next!(self, TokenType::Comma);
        }

        return Ok(Box::new(ExpressionNode::HashLiteral { params }));
    }
}

