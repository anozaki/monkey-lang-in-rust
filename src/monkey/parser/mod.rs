use std::rc::Rc;

use crate::monkey::lexer::Lexer;
use crate::monkey::parser::ast::{Program, StatementNode};
use crate::monkey::Result;
use crate::monkey::token::{Token, TokenType};
use crate::monkey::token::TokenType::Semicolon;

pub mod parser;
pub mod expression;
pub mod ast;


#[macro_export]
macro_rules! try_next {
    ($self: ident, $token: expr) => {
        if $self.current().token_type != $token {
            return Err(Error::InvalidTokenError(Rc::clone($self.current())));
        }
        $self.next();
    }
}

fn next_token(lexer: &dyn Lexer) -> Token {
    match lexer.token() {
        Ok(token) => token,
        Err(message) => panic!("Unable to parse: {message}")
    }
}

pub struct Parser<'a> {
    lexer: &'a dyn Lexer,

    token_current: Rc<Token>,
    token_peek: Rc<Token>,
}

impl <'a> Parser <'a> {
    pub fn new(lexer: &'a dyn Lexer) -> Self {
        let token_current = Rc::new(next_token(lexer));
        let token_peek = Rc::new(next_token(lexer));
        Parser {
            lexer,
            token_current,
            token_peek,
        }
    }

    pub fn current(&self) -> &Rc<Token> {
        return &self.token_current;
    }

    pub fn peek(&self) -> &Rc<Token> {
        return &self.token_peek;
    }

    pub fn next(&mut self) {
        self.token_current = Rc::clone(&self.token_peek);
        self.token_peek = Rc::new(next_token(self.lexer));
    }

    pub fn skip_semicolon(&mut self) {
        while self.current().token_type == Semicolon {
            self.next()
        }
    }

    pub fn parse_program(&mut self) -> Result<Box<Program>> {
        let mut program: Box<Program> = Box::new(Program::default());

        while self.current().token_type != TokenType::EndOfFile && self.current().token_type != TokenType::RightBrace {
            program.statements.push(self.parse_statement()?);
            self.skip_semicolon();
        }

        Ok(program)
    }

    pub fn parse_statement(&mut self) -> Result<StatementNode> {
        Ok(match self.current().token_type {
            TokenType::Let => self.parse_let()?,
            TokenType::Return => self.parse_return()?,
            TokenType::If => self.parse_if()?,
            _ => self.parse_expression()?,
        })
    }
}

#[cfg(test)]
mod test {
    use std::fs;
    use crate::monkey::lexer::MonkeyLexer;

    use super::*;

    pub fn evaluate(source: &str) -> String {
        let contents = fs::read_to_string(source).unwrap();
        let lexer = MonkeyLexer::new(&contents);

        let mut parser = Parser::new(&lexer);
        let program = parser.parse_program().unwrap();

        let mut output = String::new();
        for statement in &program.statements {
            output += &format!("{:?}\n", statement);
        }

        output
    }

    macro_rules! evaluate {
        ($name:ident, $file:literal) => {
            #[test]
            fn $name() {
                let output = evaluate(&format!("monkey/{}", $file));
                insta::assert_snapshot!(output)
            }
        }
    }

    evaluate!(test_parse_let, "test_parser_let.mky");
    evaluate!(test_parse_return, "test_parser_return.mky");
    evaluate!(test_parser_infix, "test_parser_infix.mky");
    evaluate!(test_parser_expression, "test_parser_expression.mky");
    evaluate!(test_parser_expression_prec, "test_parser_expression_prec.mky");
    evaluate!(test_parser_if, "test_parser_if.mky");
    evaluate!(test_parser_function_literals, "test_parser_function_literals.mky");
    evaluate!(test_parser_closure, "test_parser_closure.mky");
    evaluate!(test_parser_string, "test_parser_string.mky");
    evaluate!(test_parser_index, "test_parser_index.mky");
    evaluate!(test_parser_hash, "test_parser_hash.mky");
}
