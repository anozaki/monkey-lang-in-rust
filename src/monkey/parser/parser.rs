use std::rc::Rc;

use crate::try_next;
use crate::monkey::error::Error;
use crate::monkey::parser::ast::{Identifier, StatementNode};
use crate::monkey::parser::expression::OrderOfOps;
use crate::monkey::parser::Parser;
use crate::monkey::Result;
use crate::monkey::token::TokenType;

impl <'a> Parser<'a> {
    pub fn parse_let(&mut self) -> Result<StatementNode> {
        let let_token = Rc::clone(&self.current());
        self.next();

        let identifier = Identifier(self.current().literal.clone());
        self.next();

        try_next!(self, TokenType::Assign);

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        Ok(
            StatementNode::Let(
                identifier, expression,
            )
        )
    }

    pub fn parse_if(&mut self) -> Result<StatementNode> {
        let token = Rc::clone(&self.token_current);

        self.next();

        try_next!(self, TokenType::LeftParen);

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        try_next!(self, TokenType::RightParen);

        try_next!(self, TokenType::LeftBrace);

        let consequence = self.parse_program()?;

        try_next!(self, TokenType::RightBrace);

        let alternative = if self.current().token_type == TokenType::Else {
            self.next();

            try_next!(self, TokenType::LeftBrace);

            match self.parse_program() {
                Ok(result) => {
                    self.next();
                    Some(result)
                }
                Err(err) => return Err(err),
            }
        } else {
            None
        };

        Ok(
            StatementNode::If {
                condition: expression,
                consequence,
                alternative,
            }
        )
    }

    pub fn parse_return(&mut self) -> Result<StatementNode> {
        self.next();

        let expression = self.parse_expression_node(OrderOfOps::Lowest)?;

        self.next();

        Ok(StatementNode::Return(expression))
    }
}

