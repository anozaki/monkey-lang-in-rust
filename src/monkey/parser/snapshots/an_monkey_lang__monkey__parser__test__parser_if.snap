---
source: src/monkey/parser/mod.rs
expression: output
---
If { condition: Infix { operator: Equal, left: Identifier(Identifier("a")), right: Bool(true) }, consequence: Program { statements: [Let(Identifier("a"), Identifier(Identifier("b"))), If { condition: Infix { operator: Equal, left: Identifier(Identifier("b")), right: Identifier(Identifier("book")) }, consequence: Program { statements: [Let(Identifier("d"), Identifier(Identifier("e"))), Let(Identifier("d"), Infix { operator: Add, left: Infix { operator: Add, left: Infix { operator: Add, left: Identifier(Identifier("a")), right: Identifier(Identifier("b")) }, right: Identifier(Identifier("c")) }, right: Infix { operator: Add, left: Int(10), right: Int(20) } })] }, alternative: None }] }, alternative: Some(Program { statements: [Let(Identifier("a"), Infix { operator: Add, left: Identifier(Identifier("c")), right: Int(5) })] }) }

